File: frontend/src/lib/panel.ts
/**
 * The right panel's model (`docs/12` §8), derived from the transcript.
 *
 * Every view here is a function of rows the window already holds, rather than a second
 * source of truth the host has to keep in step with the conversation: the transcript *is*
 * the record of what a session did, and a file list read from anywhere else could disagree
 * with the cards on screen.
 *
 * The one thing this cannot see is a file a `bash` command wrote — the shell's own redirection
 * is not parsed, here or by the engine's reference renderer — so the Changed view says what
 * it is rather than presenting itself as a complete diff of the working tree.
 */

import type { RowSnapshot, ToolSnapshot } from "../bridge";
import { parseObject, touchedFiles, toolView } from "./toolView";
import { batchReviewPiece, orderedReviewPieces, type ReviewPiece } from "./reviewDiff";

export interface ChangedFile {
  /** Absolute whenever the engine resolved one, which it does for every write and edit. */
  path: string;
  /** `written` when any touch replaced the file's content, `edited` when only diffs did. */
  kind: "written" | "edited";
  /** How many calls touched it: "written then edited twice" is not the same change as one write. */
  touches: number;
  /**
   * The engine's diff for the call that touched it, and only when that call changed exactly
   * this file — a batched `edit` reports one joined diff for every file it touched, and
   * showing it against each of them would be a claim the engine did not make. The full
   * joined diff is on the tool card the row links to.
   */
  diff: string | null;
  /** Index into the transcript's rows: where the last touch was recorded. */
  row: number;
}

/**
 * Files this session changed, in the order it first touched them.
 *
 * First-touch order rather than most-recent-first, because the list is read while a turn
 * streams: a row that jumps position every time the agent edits it is a row nobody can click.
 */
export function changedFiles(rows: RowSnapshot[]): ChangedFile[] {
  const byPath = new Map<string, ChangedFile>();

  rows.forEach((row, index) => {
    const tool = row.tool;
    if (tool === null) return;
    // A call that has not ended has not changed anything yet: counting its arguments would
    // list a file the agent is still deciding about.
    if (!tool.finished || tool.isError) return;

    const touched = touchedFiles(tool);
    if (touched.length === 0) return;

    const diff = touched.length === 1 ? diffOf(tool) : null;

    for (const file of touched) {
      const existing = byPath.get(file.path);
      if (existing === undefined) {
        byPath.set(file.path, {
          path: file.path,
          kind: file.kind,
          touches: 1,
          diff,
          row: index,
        });
        continue;
      }

      existing.touches += 1;
      // A file the session wrote and then edited stays "written": that is the fact a
      // reviewer needs about it.
      if (file.kind === "written") existing.kind = "written";
      if (diff !== null) existing.diff = diff;
      existing.row = index;
    }
  });

  return [...byPath.values()];
}

export interface TurnChangedFile extends ChangedFile {
  /** Null when the tool gave no per-file diff (a write or a batched edit). */
  lines: { added: number; removed: number } | null;
}

export interface TurnFileSummary {
  /** Render after this transcript row, keeping search and panel row indices unchanged. */
  after: number;
  files: TurnChangedFile[];
  /** Shown only if every changed file has complete per-file counts. */
  lines: { added: number; removed: number } | null;
  /** Each successful mutating call, in order, for the Review disclosure. */
  changes: { row: number; paths: string[]; diff: string | null }[];
}

/** A request to show one turn's complete review in the right-hand Files panel. */
export interface TurnReviewRequest {
  thread: string;
  id: number;
  summary: TurnFileSummary;
}

export interface FileReview {
  /** Source-ordered hunks, with quiet markers for the untouched spans between them. */
  pieces: ReviewPiece[];
  /** Tool calls that changed the file but supplied no diff (notably whole-file writes). */
  withoutDiff: number[];
  /** A batch diff may also contain hunks for other files. */
  includesBatch: boolean;
}

export function fileReview(summary: TurnFileSummary, path: string): FileReview {
  const changes = summary.changes.filter((change) => change.paths.includes(path));
  const singleFileDiffs = changes
    .filter((change) => change.paths.length === 1)
    .map((change) => change.diff?.trimEnd())
    .filter((diff): diff is string => typeof diff === "string" && diff !== "");
  // A batch diff may contain several files. Keep that payload whole and in arrival order;
  // sorting its hunks as if they all belonged to this file would be misleading.
  const batchPieces: ReviewPiece[] = changes
    .filter((change) => change.paths.length > 1 && change.diff)
    .flatMap((change) => {
      const piece = batchReviewPiece(change.diff!);
      return piece === null ? [] : [piece];
    });

  return {
    pieces: [...orderedReviewPieces(singleFileDiffs), ...batchPieces],
    withoutDiff: changes
      .filter((change) => !change.diff || (
        change.paths.length === 1
          ? orderedReviewPieces([change.diff]).length === 0
          : batchReviewPiece(change.diff) === null
      ))
      .map((change) => change.row),
    includesBatch: changes.some((change) => change.diff !== null && change.paths.length > 1),
  };
}

/** Files changed between one user message and the next (or the settled transcript tail). */
export function turnFileSummaries(rows: RowSnapshot[], streaming: boolean): TurnFileSummary[] {
  const summaries: TurnFileSummary[] = [];
  let start = 0;

  function finish(end: number): void {
    if (end < start) return;
    const turn = rows.slice(start, end + 1);
    const files = changedFilesInTurn(turn).map((file) => ({ ...file, row: file.row + start }));
    if (files.length === 0) return;

    summaries.push({
      after: end,
      files,
      changes: turn.flatMap((row, index) => {
        const tool = row.tool;
        if (tool === null || !tool.finished || tool.isError) return [];
        const paths = touchedFiles(tool).map((file) => file.path);
        return paths.length === 0
          ? []
          : [{ row: index + start, paths, diff: diffOf(tool) }];
      }),
      lines: files.every((file) => file.lines !== null)
        ? files.reduce(
            (sum, file) => ({
              added: sum.added + (file.lines?.added ?? 0),
              removed: sum.removed + (file.lines?.removed ?? 0),
            }),
            { added: 0, removed: 0 },
          )
        : null,
    });
  }

  rows.forEach((row, index) => {
    if (row.role !== "user" || index === start) return;
    finish(index - 1);
    start = index;
  });

  // The session status, not a tool's row, decides whether the turn ended: stopped turns
  // can retain an unfinished card even though their completed edits should be summarized.
  if (!streaming) finish(rows.length - 1);
  return summaries;
}

/** Keep counts honest: a joined batch diff or a whole-file write has no per-file line total. */
function changedFilesInTurn(rows: RowSnapshot[]): TurnChangedFile[] {
  const counts = new Map<string, { added: number; removed: number } | null>();

  for (const row of rows) {
    const tool = row.tool;
    if (tool === null || !tool.finished || tool.isError) continue;
    const touched = touchedFiles(tool);
    if (touched.length === 0) continue;
    const diff = touched.length === 1 ? diffOf(tool) : null;
    const next = diff === null ? null : diffLineCounts(diff);

    for (const file of touched) {
      const previous = counts.get(file.path);
      counts.set(
        file.path,
        previous === null || next === null
          ? null
          : {
              added: (previous?.added ?? 0) + next.added,
              removed: (previous?.removed ?? 0) + next.removed,
            },
      );
    }
  }

  return changedFiles(rows).map((file) => ({ ...file, lines: counts.get(file.path) ?? null }));
}

function diffLineCounts(diff: string): { added: number; removed: number } {
  let added = 0;
  let removed = 0;
  for (const line of diff.split("\n")) {
    if (line.startsWith("+++ ") || line.startsWith("--- ")) continue;
    if (line.startsWith("+")) added += 1;
    else if (line.startsWith("-")) removed += 1;
  }
  return { added, removed };
}

/** The rows whose results spilled to an artifact, newest last. */
export interface ArtifactRef {
  /** The id in `artifact://<id>`. */
  id: string;
  tool: string;
  row: number;
}

export function artifacts(rows: RowSnapshot[]): ArtifactRef[] {
  const found: ArtifactRef[] = [];

  rows.forEach((row, index) => {
    const tool = row.tool;
    if (tool === null) return;

    const id = toolView(tool).artifact;
    if (id !== null) found.push({ id, tool: tool.toolName, row: index });
  });

  return found;
}

/**
 * Where a todo's own words appear in the transcript, if they appear at all.
 *
 * The phases arrive from `get_state`, which carries no row: the link back is the text the
 * `todo` tool was called with. A task the agent wrote and has not mentioned since still has
 * a row from the first write, which is why this looks for the *earliest* mention — that is
 * where a reader wants to be taken.
 */
export function taskRow(rows: RowSnapshot[], task: string): number | null {
  const needle = task.trim();
  if (needle === "") return null;

  for (const [index, row] of rows.entries()) {
    const tool = row.tool;
    if (tool === null || tool.toolName !== "todo") continue;
    if (
      tool.args.includes(needle) ||
      tool.details.includes(needle) ||
      tool.output.includes(needle)
    ) {
      return index;
    }
  }

  return null;
}

/** A byte count a reader can compare at a glance. */
export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${Math.round(bytes / 1024)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

/**
 * A path as a reader wants it: relative to the workspace it is inside.
 *
 * The panel shows several threads at once, each with its own workspace, so a bare absolute path
 * wastes the width that says *which* file this is. A path outside the workspace is left alone —
 * shortening it against a root it does not share would invent a relationship.
 */
export function relativeTo(path: string, root: string | null): string {
  if (root === null || root === "") return path;
  const base = root.endsWith("/") ? root : `${root}/`;
  return path.startsWith(base) ? path.slice(base.length) : path;
}

/** The engine's own unified diff for a call, when the tool reported one. */
function diffOf(tool: ToolSnapshot): string | null {
  for (const text of [tool.details, tool.args]) {
    const parsed = parseObject(text);
    const diff = parsed?.diff;
    if (typeof diff === "string" && diff !== "") return diff;
  }
  return null;
}

File: src-tauri/tests/panel.rs
//! The right panel's data, against a real engine (`docs/12` §8).
//!
//! Step 10's host half in one test: the three sources the panel reads and the one thing it
//! writes, observed end to end rather than reasoned about. What it asserts, in order:
//!
//! 1. **The plan.** `set_todos` seeds one, and what comes back is the engine's own list —
//!    compared against a *re-read* of `get_state`, because "the answer is the answer and not
//!    an echo" is only provable by asking twice. The panel's rendering of it
//!    (`ControlSnapshot.todoPhases`) is compared too. Measured here and printed: what the
//!    engine does to a phase it did not write (it projects, it does not normalise).
//! 2. **The tree.** One level of a temporary directory holding a directory and two files:
//!    directories first, cold-cased by name, sizes and timestamps measured rather than
//!    guessed — and a path outside that directory refused.
//! 3. **A real spilled artifact.** The agent is asked to run `seq 1 200000` through the
//!    `bash` tool, which is far past the engine's inline limit: the card carries a
//!    `meta.truncation` record with an id, and the file beside the session holds the whole
//!    output the card cut short. This is M5's "an artifact spills and is viewable".
//! 4. **The closed thread.** With the sidecar gone, the plan is still answered — from the
//!    control the registry remembered — and the session file a closed thread's artifacts
//!    live beside is still known.
//!
//! Isolation, because a live engine writes: the sidecar is given `--session-dir`, so every
//! session this test creates lands in a temp directory that is removed at the end, and the
//! app's `Store` is pointed at the same temp agent directory. The engine's *configuration* —
//! models, credentials, settings — stays where the user put it and is only read.
//!
//! Ignored by default: it spawns a real agent, needs provider credentials and network, and
//! costs a model call that has to actually run a command.
//!
//! ```text
//! cargo test -p omp-desktop --test panel -- --ignored --nocapture
//! ```

mod support;

use std::path::PathBuf;
use std::sync::Arc;
use std::time::Duration;

use omp_desktop::dto::RowSnapshot;
use omp_desktop::panel;
use omp_desktop::session::{open, LiveSession};
use omp_desktop::threads::Threads;
use omp_store::Store;
use omp_transport::protocol::{self, commands};
use omp_transport::{ClientOptions, SidecarSpec};
use serde_json::{json, Value};
use support::{say, RecordingSink};

/// The workspace the tree is asserted against: a directory and two files, whose contents
/// are here so the sizes the listing reports are compared with what was written.
const README: &str = "# panel\n";
const SOURCE: &str = "fn main() {}\n";

/// How many lines the command the agent is asked to run prints.
///
/// The command is `seq 1 <this>`, and the count is the point: it is far past any inline
/// limit the engine can have, so the result cannot be shown whole in a card.
const FLOOD_LINES: u64 = 200_000;

/// A flood that still spills but fits the host's read cap.
///
/// The window between the two is what this is: past the engine's inline limit (~50 KiB
/// measured on the flood above, which kept 8 709 of 200 001 lines) and under
/// [`omp_desktop::panel::MAX_ARTIFACT_BYTES`], so the artifact is handed over *whole*.
const CONTAINED_LINES: u64 = 20_000;

fn options() -> ClientOptions {
    ClientOptions {
        request_timeout: Duration::from_secs(60),
        ..Default::default()
    }
}

/// This test's own directory: the workspace, the relocated session store, and the agent
/// directory the catalogue reads, all under one path that is removed when the run ends.
fn root() -> PathBuf {
    let path = std::env::temp_dir().join(format!("omp-desktop-panel-{}", std::process::id()));
    let _ = std::fs::remove_dir_all(&path);
    path
}

/// The `seq 1 200000` output, as the file the engine spills should hold it.
///
/// Generated rather than read back, so the assertion is against what the command *must* have
/// printed: comparing the artifact with the card would prove only that they agree with each
/// other, and the card is the thing under test. `seq` writes one number per line, newline
/// terminated, and nothing else.
fn flood_output(lines: u64) -> String {
    let mut text = String::with_capacity(lines as usize * 7);
    for number in 1..=lines {
        text.push_str(&number.to_string());
        text.push('\n');
    }

    text
}

/// The truncation record a spilled tool row carries, as the card parser reads it.
///
/// The path is the one `frontend/src/lib/toolView.ts` walks (`details.meta.truncation`), and
/// the tool's own output is returned beside it: the artifact has to be *more* than this text
/// for the panel to be worth opening.
fn spill(rows: &[RowSnapshot]) -> Option<(String, Value, String)> {
    rows.iter().rev().find_map(|row| {
        let tool = row.tool.as_ref()?;
        if !tool.finished || tool.details.is_empty() {
            return None;
        }

        let details: Value = serde_json::from_str(&tool.details).ok()?;
        let record = details.get("meta")?.get("truncation")?.clone();
        let id = record.get("artifactId").and_then(Value::as_str)?;
        if id.is_empty() {
            return None;
        }

        Some((id.to_string(), record, tool.output.clone()))
    })
}

/// Ask the agent to run `seq 1 <lines>` through the `bash` tool, and answer with the
/// truncation record of the result, the card's own text, and what the command must have
/// printed.
///
/// The command is spelled out because the test's subject is the *engine's* truncation, not
/// the agent's judgement: a model that piped this into `wc -l` would produce a small result
/// and no artifact at all. Two attempts are allowed for exactly that reason — the failure
/// mode is the model's choice, not the host's behaviour.
async fn flooded(live: &LiveSession, lines: u64) -> (String, Value, String, String) {
    for attempt in 1..=2 {
        say(
            live,
            &format!(
                "Run exactly this shell command with the bash tool, with no pipes, no \
                 redirection, no `wc` and no other command: `seq 1 {lines}`. Then reply with \
                 only the last number it printed."
            ),
        )
        .await;

        let rows = live.rows().expect("the transcript is readable");
        if let Some((id, record, card)) = spill(&rows) {
            return (id, record, card, flood_output(lines));
        }

        println!("attempt {attempt}: the agent did not spill anything; asking again");
    }

    panic!("the agent must have run `seq 1 {lines}` through the bash tool: that output is far past the engine's inline limit");
}

#[tokio::test]
#[ignore = "spawns a real `omp` sidecar, needs provider credentials, and costs a model call"]
async fn the_panels_three_sources_against_a_real_engine() {
    let root = root();
    let workspace = root.join("workspace");
    let nested = workspace.join("src");
    let agent = root.join("agent");
    let sessions = agent.join("sessions/panel");
    for dir in [&workspace, &nested] {
        std::fs::create_dir_all(dir).expect("a test directory");
    }
    // Two files and a directory, which is what the level below is asserted against.
    std::fs::write(workspace.join("README.md"), README).expect("a file");
    std::fs::write(nested.join("main.rs"), SOURCE).expect("a file");

    let store = Store::at(&agent);
    let threads = Arc::new(Threads::default());
    let spec = SidecarSpec::omp(&workspace)
        // Every session this run creates lands here, and is removed with the directory.
        .with_arg("--session-dir", sessions.to_str().expect("a utf-8 path"))
        // This test is about the panel's data, not the gate in front of it: the app's own
        // launch is `--approval-mode write`, which asks before an exec, and an unanswered
        // dialog would stall the turn this test needs to finish. `approval.rs` is where the
        // asking is asserted.
        .with_approval_mode("yolo");

    let live = open(&spec, options(), RecordingSink::default(), threads.clone())
        .await
        .expect("the sidecar opens");
    threads.insert(Arc::clone(&live));
    let id = live.thread.current();
    let session_file = live.session_file().expect("the session is persisted");
    println!("session: {id}");
    println!("session file: {session_file}");

    // When the catalogue can see a *live* session, which is not the same question as
    // whether the session exists: the engine owns when the file lands. Printed rather than
    // asserted, because what the host needs is the answer for a *closed* thread (asserted
    // below) and this is the measurement that says the two are not the same thing.
    let mut waited = 0;
    while store.find(&id).is_none() && waited < 30 {
        waited += 1;
        tokio::time::sleep(Duration::from_millis(100)).await;
    }
    println!(
        "the catalogue {} the live session at the handshake",
        match waited {
            0 => "already listed".to_string(),
            waited if waited >= 30 => "had not listed after 3 s of polling".to_string(),
            waited => format!("listed after {} ms of polling", waited * 100),
        }
    );

    // ---- 1. The plan ------------------------------------------------------
    // The write half of `set_todos`: `panel::outgoing` is the host's only rule (an empty
    // list is refused), and what is asserted here is the engine's answer to the rest.
    let seeded = panel::outgoing(&[omp_desktop::dto::TodoPhaseInput {
        name: "Measurement".to_string(),
        tasks: vec![
            omp_desktop::dto::TodoTaskInput {
                content: "seed the plan".to_string(),
                status: "completed".to_string(),
                blocker: None,
            },
            omp_desktop::dto::TodoTaskInput {
                content: "read it back".to_string(),
                status: "in_progress".to_string(),
                blocker: None,
            },
        ],
    }])
    .expect("a plan with a phase in it is sent");

    let stored = live
        .set_todo_phases(seeded)
        .await
        .expect("the engine stores the plan");
    println!("set_todos answered: {stored:?}");
    assert_eq!(stored.len(), 1);
    assert_eq!(stored[0].name, "Measurement");
    assert_eq!(stored[0].tasks.len(), 2);
    assert_eq!(stored[0].tasks[0].content, "seed the plan");
    assert_eq!(stored[0].tasks[1].status, "in_progress");

    // The answer is only worth rendering if it is what the engine *holds*, so ask again by
    // the other road: `thread_todos` re-reads `get_state`, and it must agree.
    let reread = live.todo_phases().await;
    assert_eq!(
        panel::phases(&reread),
        panel::phases(&stored),
        "a re-read must agree with the answer, or the panel is rendering a copy"
    );

    // And the control snapshot the window renders carries the same list, whole.
    let control = live.status().expect("status").control;
    assert_eq!(control.todo_phases, panel::phases(&stored));

    // What the engine does with a plan it did not write. Measured rather than assumed, and
    // printed: it stores what it is given (no normalisation) and answers with its own
    // projection of it, so a task can come back *without* the status it was sent.
    let client = live.client().expect("the session is open");
    let response = client
        .request(
            commands::set_todos(json!([{
                "name": "Raw",
                "tasks": [
                    { "content": "no status", "id": 7, "note": "extra" },
                    { "content": "with a blocker", "status": "blocked", "blocker": "the review" },
                ],
            }])),
            None,
        )
        .await
        .expect("the engine answers set_todos");
    assert!(protocol::is_success(&response), "refused: {response}");
    let echoed = &response["data"]["todoPhases"];
    println!("the engine's answer to a raw plan: {echoed}");
    assert_eq!(echoed[0]["name"], json!("Raw"));
    let task = echoed[0]["tasks"][0]
        .as_object()
        .expect("a task comes back as an object");
    assert_eq!(task.get("content"), Some(&json!("no status")));
    assert!(
        !task.contains_key("id") && !task.contains_key("note"),
        "the engine answers with its own projection and drops what it does not know: {task:?}"
    );
    assert_eq!(echoed[0]["tasks"][1]["blocker"], json!("the review"));
    println!("a task sent with no status came back as {task:?}");

    // Put a real plan back, which is also what the next section reads off the control.
    let restored = live
        .set_todo_phases(json!([{
            "name": "Measurement",
            "tasks": [{ "content": "read it back", "status": "in_progress" }],
        }]))
        .await
        .expect("the plan is restored");
    assert_eq!(restored[0].tasks.len(), 1);

    // ---- 2. The tree ------------------------------------------------------
    let level = panel::tree(&live.workspace, None).expect("the workspace lists");
    println!(
        "workspace level: {:?}",
        level.iter().map(|entry| &entry.name).collect::<Vec<_>>()
    );
    let names: Vec<&str> = level.iter().map(|entry| entry.name.as_str()).collect();
    assert_eq!(
        names,
        vec!["src", "README.md"],
        "directories first, then the files, case-insensitively by name"
    );

    let directory = &level[0];
    assert!(directory.is_dir);
    assert_eq!(directory.size, 0, "a directory has no size to draw");
    assert!(
        directory.path.starts_with(&live.workspace),
        "a row's path is inside the workspace: {}",
        directory.path
    );

    let file = &level[1];
    assert!(!file.is_dir);
    assert_eq!(file.size, README.len() as u64, "the size is the file's own");
    assert!(file.modified_at > 0, "a real file has a timestamp");

    // The path a row carries is the path the next level is asked for — the panel passes it
    // straight back, so a listing that could not be re-listed would be a dead row.
    let below = panel::tree(&live.workspace, Some(&directory.path)).expect("the next level");
    assert_eq!(below.len(), 1);
    assert_eq!(below[0].name, "main.rs");
    assert_eq!(below[0].size, SOURCE.len() as u64);

    // Everything outside the thread's workspace is refused, including the directory the
    // workspace sits in — that is the whole of the containment rule.
    let outside = &root.to_string_lossy().to_string();
    let refused = panel::tree(&live.workspace, Some(outside))
        .expect_err("a path outside the workspace is refused");
    println!("refused: {refused}");
    assert!(
        refused.contains("outside the thread's workspace"),
        "{refused}"
    );

    // ---- 3. A real spilled artifact ---------------------------------------
    // The id names a file beside the session's own `.jsonl`, which is the engine's rule and
    // the whole reason the host can find it without an RPC command.
    let directory = panel::artifacts_dir(&session_file).expect("the engine's rule");

    let (artifact_id, record, card_text, expected) = flooded(&live, FLOOD_LINES).await;
    println!("truncation record: {record}");
    println!("the card kept {} bytes of text", card_text.len());

    let read = panel::artifact(&directory, &artifact_id).expect("the artifact is read");
    println!(
        "artifact {} : {} bytes at {} (truncated in the read: {})",
        read.id, read.bytes, read.path, read.truncated
    );
    println!("the whole result is {} bytes", expected.len());

    assert_eq!(read.id, artifact_id);
    assert!(
        std::path::Path::new(&read.path).starts_with(&directory),
        "the file read is inside the session's artifacts directory: {}",
        read.path
    );
    // The artifact is the *complete* output, which is what the card could not show. The
    // record's own account of the stream is the engine's, and it must match the file.
    assert_eq!(
        read.bytes,
        expected.len() as u64,
        "the spilled file *is* the whole output the card cut short: byte for byte, not a \
         summary of it"
    );
    if let Some(total) = record.get("totalBytes").and_then(Value::as_u64) {
        assert_eq!(
            total, read.bytes,
            "the truncation record's own account of the stream is the file it spilled to"
        );
    }
    assert!(
        (card_text.len() as u64) < read.bytes,
        "the card was truncated, the artifact is not: {} bytes of text against {} on disk",
        card_text.len(),
        read.bytes
    );
    // The head of the file is the command's own first lines: a summary or a tail would not
    // start here. (The read is capped at 256 KiB, so a 1.3 MB artifact arrives truncated —
    // and this is the assertion that says the *start* of what arrived is the real output.)
    assert!(
        read.text.starts_with("1\n2\n3\n"),
        "the artifact begins with the command's output, got {:?}",
        read.text.chars().take(40).collect::<String>()
    );
    assert!(
        expected.starts_with(&read.text),
        "everything the read returned is the command's own output, in order ({} bytes read)",
        read.text.len()
    );
    println!(
        "the artifact is the command's whole output: {} bytes on disk, {} generated",
        read.bytes,
        expected.len()
    );
    if read.truncated {
        assert_eq!(
            read.text.len(),
            panel::MAX_ARTIFACT_BYTES as usize,
            "a capped read stops exactly at the cap"
        );
    }
    println!(
        "direction {:?}, truncatedBy {:?}, totalLines {:?}, outputLines {:?}, elidedBytes {:?}",
        record.get("direction"),
        record.get("truncatedBy"),
        record.get("totalLines"),
        record.get("outputLines"),
        record.get("elidedBytes"),
    );

    // ---- 4. An artifact the host can hand over whole -----------------------
    // The engine spilled this one too — the card was cut short — but the file fits the
    // host's cap, so what the panel shows *is* the whole result: the same claim the section
    // above makes about the file's size, made about its text.
    let (contained_id, record, card_text, expected) = flooded(&live, CONTAINED_LINES).await;
    println!("contained artifact {contained_id}: record {record}");

    let contained = panel::artifact(&directory, &contained_id).expect("the artifact is read");
    println!(
        "{} bytes on disk, {} of text, truncated in the read: {}",
        contained.bytes,
        contained.text.len(),
        contained.truncated
    );
    assert_eq!(contained.bytes, expected.len() as u64);
    assert!(
        !contained.truncated,
        "the whole file is inside the cap, so the flag must not say otherwise"
    );
    assert_eq!(
        contained.text, expected,
        "the text is the command's output, character for character"
    );
    assert!(
        (card_text.len() as u64) < contained.bytes,
        "the card was still the truncated one: {} bytes of text against {} on disk",
        card_text.len(),
        contained.bytes
    );

    // ---- 5. The closed thread --------------------------------------------
    // The same question once the session has done some work: printed, because the two
    // answers are the point — a catalogue scan is a question about the engine's flush
    // schedule while a session is running, and about the file itself once it has closed.
    println!(
        "the catalogue lists the live session after two turns: {}",
        store.find(&id).is_some()
    );

    let live_plan = live.status().expect("status").control.todo_phases;
    threads.remove(&id).expect("the thread is registered");
    live.shutdown(Duration::from_secs(10)).await;

    // The session's own record of where it worked, which is what the tree is rooted at once
    // the sidecar is gone (and the only thing that outlives it). Read after the shutdown
    // rather than before it: a catalogue scan of a *running* session is a scan of a file the
    // engine is still deciding when to write.
    let catalogued = store
        .find(&id)
        .expect("the catalogue lists the session once it has closed");
    assert_eq!(
        catalogued.cwd, live.workspace,
        "the header's cwd is the workspace the panel browses"
    );

    let cached = threads
        .cached_control(&id)
        .expect("a closed thread is remembered, or the panel would go blank");
    assert_eq!(
        panel::phases(&cached.todo_phases),
        live_plan,
        "the plan a closed thread answers with is the one the engine last held"
    );
    assert_eq!(
        cached.session_file.as_deref(),
        Some(session_file.as_str()),
        "and its artifacts are still resolvable through the remembered session file"
    );
    assert!(
        threads.cached_control("01never-opened").is_none(),
        "a thread this window never opened has no plan to answer with"
    );

    // `read_artifact` for a closed thread resolves the same directory: nothing about the
    // read depended on the sidecar being alive.
    let again = panel::artifact(&directory, &artifact_id).expect("the artifact still reads");
    assert_eq!(again, read);

    let _ = std::fs::remove_dir_all(&root);
    println!(
        "done: plan, tree and artifact {} all read back",
        artifact_id
    );
}
